modules/core/04-channel/keeper/keeper_test.go

```diff
@@ -541,3 +542,321 @@ func (suite *KeeperTestSuite) TestUnsetParams() {
+					suite.chainA.GetContext(),
+					path.EndpointA.ChannelConfig.PortID,
+					path.EndpointA.ChannelID,
+					4, // limit == 4
+				)
+				suite.Require().NoError(err)
+
+				// We expect 4 to be pruned and 6 left.
+				suite.Require().Equal(uint64(4), pruned)
+				suite.Require().Equal(uint64(6), left)
+
+				// Check state post-prune
+				postPruneExpState(6, 6, 5)
+
+				// Previous upgrade is complete, send additional packets and do yet another upgrade.
+				// This is _after_ the first upgrade.
+				suite.sendMockPackets(path, 10)
+
+				// Do another upgrade.
+				upgradeFields = types.UpgradeFields{Version: fmt.Sprintf("%s-v3", ibcmock.Version)}
+				suite.UpgradeChannel(path, upgradeFields)
+
+				// A total of 16 stale acks/receipts exist on A. Prune 10 of them (default in test).
+			},
+			func(pruned, left uint64) {
+				// Expected state should be 6 acks/receipts left, sequenceStart == 15.
+				postPruneExpState(6, 6, 15)
+
+				// We expect 10 to be pruned and 6 left.
+				suite.Require().Equal(uint64(10), pruned)
+				suite.Require().Equal(uint64(6), left)
+			},
+			nil,
+		},
+		{
+			"success: unordered -> ordered -> unordered, acksLen != receiptsLen after packet sends",
+			func() {
+				// Send 5 packets from B -> A, creating 5 packet receipts and 5 packet acks on A.
+				// This is _before_ the first upgrade.
+				suite.sendMockPackets(path, 5)
+
+				// Set Order for upgrade to Ordered.
+				upgradeFields = types.UpgradeFields{Version: fmt.Sprintf("%s-v2", ibcmock.Version), Ordering: types.ORDERED}
+			},
+			func() {
+				// Previous upgrade is complete, send additional packets now on ordered channel (only acks!)
+				suite.sendMockPackets(path, 10)
+
+				// Do another upgrade (go back to Unordered)
+				upgradeFields = types.UpgradeFields{Version: fmt.Sprintf("%s-v3", ibcmock.Version), Ordering: types.UNORDERED}
+				suite.UpgradeChannel(path, upgradeFields)
+			},
+			func(pruned, left uint64) {
+				// After pruning 10 sequences we should be left with 5 acks and zero receipts.
+				postPruneExpState(5, 0, 11)
+
+				// We expect 10 to be pruned and 5 left.
+				suite.Require().Equal(uint64(10), pruned)
+				suite.Require().Equal(uint64(5), left)
+			},
+			nil,
+		},
+		{
+			"success: packets sent before upgrade are pruned, after upgrade are not",
+			func() {
+				// Send 5 packets from B -> A, creating 5 packet receipts and 5 packet acks on A.
+				suite.sendMockPackets(path, 5)
+			},
+			func() {},
+			func(pruned, left uint64) {
+				// We expect 5 to be pruned and 0 left.
+				suite.Require().Equal(uint64(5), pruned)
+				suite.Require().Equal(uint64(0), left)
+
+				// channel upgraded, send additional packets and try and prune.
+				suite.sendMockPackets(path, 12)
+
+				// attempt to prune 5.
+				pruned, left, err := suite.chainA.App.GetIBCKeeper().ChannelKeeper.PruneAcknowledgements(
+					suite.chainA.GetContext(),
+					path.EndpointA.ChannelConfig.PortID,
+					path.EndpointA.ChannelID,
+					5,
```

Comment:
```suggestion
					17,
```
should we be attempting to prune post upgrade packet sends by setting it to a higher limit? To show that it safely returns after all possible have been pruned?